go/sloppy/sloppy_test.go

```diff
@@ -0,0 +1,546 @@
+	enc     []interface{}
+}
+
+func (tt *testSloppyEncoder) emitLiteral(b byte) bool {
+	tt.current = append(tt.current, b)
+	return false
+}
+
+func (tt *testSloppyEncoder) emitCopy(offset, length uint16) bool {
+	tt.emitString()
+	tt.enc = append(tt.enc, offset, length)
+	return false
+}
+
+func (tt *testSloppyEncoder) emitString() {
+	if len(tt.current) > 0 {
+		tt.enc = append(tt.enc, string(tt.current))
+		tt.current = make([]byte, 0)
+	}
+}
+
+func TestCopyLongEnough(t *testing.T) {
+	assert.False(t, copyLongEnough(1, 3))
+	assert.True(t, copyLongEnough(1, 4))
+	assert.False(t, copyLongEnough(2047, 3))
+	assert.True(t, copyLongEnough(2048, 4))
+	assert.False(t, copyLongEnough(2048+1, 7))
+	assert.True(t, copyLongEnough(2048+1, 8))
+	assert.False(t, copyLongEnough(2048+1024+1, 15))
+	assert.True(t, copyLongEnough(2048+1024+1, 16))
+	assert.False(t, copyLongEnough(2048+1024+512+1, 31))
+	assert.True(t, copyLongEnough(2048+1024+512+1, 32))
+	assert.False(t, copyLongEnough(2048+1024+512+256+1, 63))
+	assert.True(t, copyLongEnough(2048+1024+512+256+1, 64))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+1, 127))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+1, 128))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+64+1, 255))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+64+1, 256))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+64+32+1, 511))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+64+32+1, 512))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+64+32+16+1, 1023))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+64+32+16+1, 1024))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+64+32+16+8+1, 2047))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+64+32+16+8+1, 2048))
+	assert.False(t, copyLongEnough(2048+1024+512+256+128+64+32+16+8+4+1, 4095))
+	assert.True(t, copyLongEnough(2048+1024+512+256+128+64+32+16+8+4+1, 4096))
+}
+
+func TestSloppySimple(t *testing.T) {
+	mo := maxOffset
+	defer func() {
+		maxOffset = mo
+	}()
+
+	// NOTE: Sloppy always buffers the last three bytes of input.
+	tc := []struct {
+		max int
+		pt  string
+		enc []interface{}
+	}{
+		// Match length not long enough
+		{4, "ababxxxx", []interface{}{"ababx"}},
+		// Trailing literal
+		{4, "01230123xxxx", []interface{}{"0123", uint16(4), uint16(4), "x"}},
+		// Match past current
+		{4, "012301230123xxxx", []interface{}{"0123", uint16(4), uint16(8), "x"}},
+		// Offset to most recent
+		{6, "ABCDxABCDyABCDzzzz", []interface{}{"ABCDx", uint16(5), uint16(4), "y", uint16(5), uint16(4), "z"}},
+		// Offset beyond max
+		{5, "0123x0123yy0123zzzz", []interface{}{"0123x", uint16(5), uint16(4), "yy0123z"}},
+		// Prefer closer match
+		{128, "23hello my friend, 12hello my friend, 01hello my friend, 23hello my friendxxxx",
+			[]interface{}{"23hello my friend, 12", uint16(19), uint16(17),
+				"01", uint16(19), uint16(17),
+				"23", uint16(19), uint16(15),
+				"x",
+			}},
+	}
+
+	for _, c := range tc {
```

Comment:
Use local test function instead. Table driven tests don't tell you which one fails.

Reply:
done